google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gct.idea.appengine.cloud;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.Files;
+import com.google.gct.idea.util.GctBundle;
+import com.google.gct.login.CredentialedUser;
+import com.google.gct.login.Services;
+import com.google.gdt.eclipse.login.common.GoogleLoginState;
+import com.google.gson.Gson;
+
+import com.intellij.execution.ExecutionException;
+import com.intellij.execution.configurations.GeneralCommandLine;
+import com.intellij.execution.configurations.GeneralCommandLine.ParentEnvironmentType;
+import com.intellij.execution.process.OSProcessHandler;
+import com.intellij.execution.process.ProcessHandler;
+import com.intellij.execution.process.ProcessListener;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.remoteServer.runtime.RemoteOperationCallback;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.Map;
+
+/**
+ * Base class for App Engine runnable actions - e.g. deploy, stop. Provides implementations
+ * for executing CLI based commands.
+ */
+public abstract class AppEngineAction implements Runnable {
+  private static final Logger logger = Logger.getInstance(AppEngineAction.class);
+
+  private LoggingHandler loggingHandler;
+  private AppEngineHelper appEngineHelper;
+  private File credentialsPath;
+
+  public AppEngineAction(
+      @NotNull LoggingHandler loggingHandler,
+      @NotNull AppEngineHelper appEngineHelper,
+      @NotNull RemoteOperationCallback callback) {
+    this.loggingHandler = loggingHandler;
+    this.appEngineHelper = appEngineHelper;
+
+    credentialsPath = createApplicationDefaultCredentials();
```

Comment:
we probably want to minimize the window between when we create the credential file and we execute.  Could we put this this in executeProvess?